File: Nejman.Newtone.Core/Data/PlaylistsAction.cs
using Nejman.Newtone.Core.Implementations;
using Nejman.Newtone.Core.IO;
using Nejman.Newtone.Core.Media;
using Nejman.Newtone.Core.Models;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

using Local = Nejman.Newtone.Core.Localization.Localization;

namespace Nejman.Newtone.Core.Data
{
    public static class PlaylistsAction
    {
        public static void Add(string playlistName, List<MediaSource> tracks)
        {
            foreach(var track in tracks)
            {
                Playlists.Add(playlistName, track);
            }
            DataFile.Save();
        }

        public static void Add(string playlistName, List<string> tracks, bool save = true)
        {
            foreach (var track in tracks)
            {
                Playlists.Add(playlistName, track);
            }

            if(save)
            {
                DataFile.Save();
            }
        }

        public static async Task Add(List<MediaSource> tracks)
        {
            var selected = await SelectPlaylist();

            if (selected == null)
            {
                return;
            }

            Add(selected, tracks);
            SnackbarImplementation.Current.Show(Local.Ready);
        }

        public static async Task Add(List<string> tracks)
        {
            var selected = await SelectPlaylist();

            if (selected == null)
            {
                return;
            }

            Add(selected, tracks);
            SnackbarImplementation.Current.Show(Local.Ready);
        }
        public static void Add(string playlistName, MediaSource track)
        {
            Add(playlistName, track.Path);
        }
        public static void Add(string playlistName, string track)
        {
            Playlists.Add(playlistName, track);
        }
        public static async Task Add(MediaSource track)
        {
            await Add(track.Path);
        }

        public static async Task Add(string track)
        {
            var selected = await SelectPlaylist();

            if(selected == null)
            {
                return;
            }

            Playlists.Add(selected, track);
            SnackbarImplementation.Current.Show(Local.Ready);
        }

        public static async Task Remove(string playlistName)
        {
            if(!CoreGlobal.Playlists.ContainsKey(playlistName))
            {
                return;
            }

            var selected = await MessageImplementation.Current.Show(Local.QuestionDeletePlaylist, Local.QuestionDeletePlaylist + " " + playlistName + "?", Local.Yes, Local.No);

            if(selected == Local.No || selected == null)
            {
                return;
            }

            Playlists.Remove(playlistName);
            SnackbarImplementation.Current.Show(Local.Ready);
        }

        public static async Task Remove(string playlistName, MediaSource track)
        {
            await Remove(playlistName, track.Path);
        }

#pragma warning disable CS1998 // Metoda asynchroniczna nie zawiera operatorów „await” i zostanie uruchomiona synchronicznie
        public static async Task Remove(string playlistName, string track)
#pragma warning restore CS1998 // Metoda asynchroniczna nie zawiera operatorów „await” i zostanie uruchomiona synchronicznie
        {
            if (!CoreGlobal.Playlists.ContainsKey(playlistName))
            {
                return;
            }

            Playlists.Remove(playlistName, track);
            SnackbarImplementation.Current.Show(Local.Ready);
        }

        public static List<PlaylistModel> GetPlaylists()
        {
            List<PlaylistModel> models = new List<PlaylistModel>();

            foreach(var name in CoreGlobal.Playlists.Keys)
            {
                var playlist = CoreGlobal.Playlists[name];

                if(playlist.Count > 0)
                {
                    var withImage = playlist.FirstOrDefault(trackPath =>
                    {
                        if(!CoreGlobal.Audios.ContainsKey(trackPath))
                        {
                            return false;
                        }

                        var track = CoreGlobal.Audios[trackPath];

                        return track.Image != null && track.Image.Length > 0;
                    });

                    var image = withImage != null && CoreGlobal.Audios.ContainsKey(withImage) ? CoreGlobal.Audios[withImage].Image : new byte[0];

                    models.Add(new PlaylistModel(name, image));
                }
            }

            return models;
        }

        public static List<PlaylistItemModel> GetPlaylist(string playlistName)
        {
            List<PlaylistItemModel> models = new List<PlaylistItemModel>();

            if(CoreGlobal.Playlists.ContainsKey(playlistName))
            {
                foreach(var trackPath in CoreGlobal.Playlists[playlistName])
                {
                    var source = TracksAction.Get(trackPath);

                    if(source != null)
                    {
                        models.Add(new PlaylistItemModel(source));
                    }
                }
            }

            return models;
        }

        public static List<PlaylistItemModel> GetPlaylistInvariant(string invariantPlaylistName)
        {
            string playlistName = null;

            foreach(var key in CoreGlobal.Playlists.Keys)
            {
                if(key.ToLowerInvariant().Contains(invariantPlaylistName.ToLowerInvariant()))
                {
                    playlistName = key;
                    break;
                }
            }

            if(playlistName == null)
            {
                return new List<PlaylistItemModel>();
            }

            return GetPlaylist(playlistName);
        }

        public static List<MediaSource> GetCurrentPlaylist()
        {
            return CoreGlobal.CurrentPlaylist;
        }

        public static async Task ChangeName(string playlistName)
        {
            var newName = await MessageImplementation.Current.ShowPrompt(Local.ChangeName, playlistName, "OK", Local.Cancel, Local.NewPlaylistHint, playlistName);

            if(CoreGlobal.Playlists.ContainsKey(newName))
            {
                await MessageImplementation.Current.Show(Local.ChangeName, Local.PlaylistExists, "OK", Local.Cancel);
                return;
            }

            Playlists.ChangeName(playlistName, newName);
            SnackbarImplementation.Current.Show(Local.Ready);
        }

        internal static async Task<string> SelectPlaylist(string newPlaylistDefaultName = "")
        {
            List<string> positions = new List<string>() { Local.NewPlaylist };

            foreach (var playlist in CoreGlobal.Playlists.Keys)
            {
                positions.Add(playlist);
            }

            var selected = await MessageImplementation.Current.Show(Local.ChoosePlaylist, Local.Cancel, positions.ToArray());

            if(selected == Local.Cancel)
            {
                return null;
            }

            if(selected == Local.NewPlaylist)
            {
                var newPlaylistName = await MessageImplementation.Current.ShowPrompt(Local.NewPlaylist, Local.NewPlaylistHint, Local.Add, Local.Cancel, Local.NewPlaylist, newPlaylistDefaultName == "" ? Local.NewPlaylist : newPlaylistDefaultName);

                if(newPlaylistName == Local.Cancel || newPlaylistName == null)
                {
                    return null;
                }

                Playlists.CreateIfNotExists(newPlaylistName);
                selected = newPlaylistName;
            }

            return selected;
        }
    }
}

File: app/src/main/java/com/nejman/nsec/music_player/core/data/PlaylistsAction.java
package com.nejman.nsec.music_player.core.data;

import com.nejman.nsec.music_player.MainActivity;
import com.nejman.nsec.music_player.R;
import com.nejman.nsec.music_player.core.DataContainer;
import com.nejman.nsec.music_player.core.loaders.DataLoader;
import com.nejman.nsec.music_player.core.models.PlaylistModel;
import com.nejman.nsec.music_player.ui.AlertDialogFragment;
import com.nejman.nsec.music_player.ui.ListDialogFragment;
import com.nejman.nsec.music_player.ui.PromptDialogFragment;

import java.util.ArrayList;
import java.util.List;
import java.util.function.Consumer;

public class PlaylistsAction {
    public static void add(String playlistName, List<String> tracks) {
        add(playlistName, tracks, true);
    }

    public static void add(String playlistName, List<String> tracks, boolean save) {
        for (String track : tracks) {
            Playlists.add(playlistName, track);
        }

        if (save) {
            try {
                DataLoader.save();
            } catch (Throwable throwable) {
                throwable.printStackTrace();
            }
        }
    }

    public static void add(List<String> tracks) {
        selectPlaylist(selected -> {
            add(selected, tracks);
            MainActivity.toast(R.string.ready);
        });
    }

    public static void add(String track) {
        selectPlaylist(selected -> {
            Playlists.add(selected, track);
            try {
                DataLoader.save();
            } catch (Throwable throwable) {
                throwable.printStackTrace();
            }
            MainActivity.toast(R.string.ready);
        });
    }

    public static void remove(String playlistName) {
        if (!DataContainer.getInstance().getPlaylists().exists(playlistName)) {
            return;
        }

        AlertDialogFragment alert = new AlertDialogFragment(MainActivity.getResString(R.string.question_delete_playlist), MainActivity.getResString(R.string.question_delete_playlist) + " " + playlistName + "?", MainActivity.getResString(R.string.yes), MainActivity.getResString(R.string.no), selected -> {
            if (!selected) {
                return;
            }
            Playlists.remove(playlistName);
            MainActivity.toast(R.string.ready);
        });
        alert.show(MainActivity.instance.getSupportFragmentManager(), "alert");
    }

    public static void remove(String playlistName, String track) {
        if (!DataContainer.getInstance().getPlaylists().exists(playlistName)) {
            return;
        }

        Playlists.remove(playlistName, track);
        MainActivity.toast(R.string.ready);
    }

    public static void changeName(String playlistName) {
        PromptDialogFragment prompt = new PromptDialogFragment(MainActivity.getResString(R.string.change_name), playlistName, "OK", MainActivity.getResString(R.string.cancel), playlistName, newName -> {
            if (DataContainer.getInstance().getPlaylists().exists(newName)) {
                MainActivity.toast(R.string.playlist_exists);
                return;
            }

            Playlists.changeName(playlistName, newName);
            MainActivity.toast(R.string.ready);
        });

        prompt.show(MainActivity.instance.getSupportFragmentManager(), "prompt");
    }

    private static void selectPlaylist(Consumer<? super String> selected) {
        selectPlaylist("", selected);
    }

    public static void selectPlaylist(String newPlatylistDefaultName, Consumer<? super String> selected) {
        List<String> positions = new ArrayList<>();
        positions.add(MainActivity.getResString(R.string.new_playlist));

        for (PlaylistModel playlist : DataContainer.getInstance().getPlaylists().getAll()) {
            positions.add(playlist.name);
        }

        ListDialogFragment dialog = new ListDialogFragment(MainActivity.getResString(R.string.choose_playlist), positions, selectedInternal -> {
            if (selectedInternal.equals(MainActivity.getResString(R.string.new_playlist))) {
                PromptDialogFragment prompt = new PromptDialogFragment(MainActivity.getResString(R.string.new_playlist), MainActivity.getResString(R.string.new_playlist_hint), MainActivity.getResString(R.string.add), MainActivity.getResString(R.string.cancel), newPlatylistDefaultName.equals("") ? MainActivity.getResString(R.string.new_playlist) : newPlatylistDefaultName, selectedPlaylist -> {
                    Playlists.createIfNotExists(selectedPlaylist);
                    selected.accept(selectedPlaylist);
                });
                prompt.show(MainActivity.instance.getSupportFragmentManager(), "prompt");
            } else {
                selected.accept(selectedInternal);
            }
        });
        dialog.show(MainActivity.instance.getSupportFragmentManager(), "dialog");
    }
}
